### PR TITLE
ROI maintenance

### PR DESCRIPTION
**Pull request by KapRom**
Feature allows to:
- read roi from png file to an object
- create png roi file from an object
- list all roi from folder

Contains Roi representation.

### CODE REVIEW

**Comment by gmrukwa** [comment]
[ROI storage & maintenance](https://trello.com/c/qG2J253q/276-roi-storage-maintenance)

File: src/Spectre.Data/Datasets/IRoiDictionary.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        /// Removes the record with specified name.
+        /// </summary>
+        /// <param name="name">The name.</param>
+        void Remove(string name);
+
+        /// <summary>
+        /// Gets the roi names.
+        /// </summary>
+        /// <returns>Names of all rois in dictionary.</returns>
+        List<string> GetRoiNames();
```
Comment (Daroslaw) [inline src/Spectre.Data/Datasets/IRoiDictionary.cs]: Maybe `IList`?

File: src/Spectre.Data/RoiIo/RoiConverter.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                (uint)bitmap.Width,
+                (uint)bitmap.Height,
+                new List<RoiPixel>());
+
+            for (int xcoordinate = 0; xcoordinate < bitmap.Width; xcoordinate++)
+            {
+                for (int ycoordinate = 0; ycoordinate < bitmap.Height; ycoordinate++)
+                {
+                    color = bitmap.GetPixel(xcoordinate, ycoordinate);
+                    if (color.B == roiColor)
```
Comment (Daroslaw) [inline src/Spectre.Data/RoiIo/RoiConverter.cs:55]: Can't we just compare it to `Color.Black`? It would be more intuitive and less tolerant for other colors.

File: src/Spectre.Data/RoiIo/RoiReader.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        public List<string> GetAllNamesFromDirectory(string path)
+        {
+            var names = new List<string>();
+
+            var allfiles = System.IO.Directory.GetFiles(path, "*.png", SearchOption.TopDirectoryOnly);
+
+            names = allfiles.ToList();
+            names.Sort();
+
+            if (names.Count == 0)
```
Comment (Daroslaw) [inline src/Spectre.Data/RoiIo/RoiReader.cs]: It would be IMO better to perform this check before conversion to list and sorting.

File: src/Spectre.Data/RoiIo/RoiReader.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        /// <param name="path">The path.</param>
+        /// <returns>
+        /// Returns names of all files in specified directory.
+        /// </returns>
+        /// <exception cref="FileNotFoundException">No *.png files found in directory.</exception>
+        public List<string> GetAllNamesFromDirectory(string path)
+        {
+            var names = new List<string>();
+
+            var allfiles = System.IO.Directory.GetFiles(path, "*.png", SearchOption.TopDirectoryOnly);
```
Comment (Daroslaw) [inline src/Spectre.Data/RoiIo/RoiReader.cs]: Shouldn't we use dependency injection for filesystem access? I wonder how the PNG's would be mocked tho

**Review by Daroslaw — CHANGES_REQUESTED** [review]
Only minor changes, overally great job!

File: src/Spectre.Data/RoiIo/RoiConverter.cs (hunk 2 above)
Comment (KapRom) [reply src/Spectre.Data/RoiIo/RoiConverter.cs:55]: That's the way gmrukwa requested

Comment (gmrukwa) [reply src/Spectre.Data/RoiIo/RoiConverter.cs:55]: @Daroslaw, I requested naming the constant in the way it is actually used here. Therefore, since we're looking for ROI, we check for `roiColor`, regardless the true value of its color.

File: src/Spectre.Data.Tests/RoiReaderTests.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+            Assert.AreEqual(actual: allRoisFromDirectory.Count, expected: DataStub.ExpectedNumberOfRoisInDirectory, message: "Incorrect number of rois read from directory");
+
+            Assert.AreEqual(actual: allRoisFromDirectory[0].Name, expected: DataStub.AddRoiDataset.Name);
+            Assert.AreEqual(actual: allRoisFromDirectory[1].Name, expected: DataStub.ReadRoiDataset.Name);
+            Assert.AreEqual(actual: allRoisFromDirectory[2].Name, expected: DataStub.WriteRoiDataset.Name);
+
+            Assert.AreEqual(actual: allRoisFromDirectory[0].Height, expected: DataStub.AddRoiDataset.Height);
+            Assert.AreEqual(actual: allRoisFromDirectory[1].Height, expected: DataStub.ReadRoiDataset.Height);
+            Assert.AreEqual(actual: allRoisFromDirectory[2].Name, expected: DataStub.WriteRoiDataset.Name);
```
Comment (gmrukwa) [inline src/Spectre.Data.Tests/RoiReaderTests.cs]: Here we probably should compare `Height`. Or we just want to be *really* sure, that the name is the same. :)

File: src/Spectre.Data.Tests/RoiReaderTests.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+            Assert.AreEqual(actual: allRoisFromDirectory[0].Width, expected: DataStub.AddRoiDataset.Width);
+            Assert.AreEqual(actual: allRoisFromDirectory[1].Width, expected: DataStub.ReadRoiDataset.Width);
+            Assert.AreEqual(actual: allRoisFromDirectory[2].Name, expected: DataStub.WriteRoiDataset.Name);
```
Comment (gmrukwa) [inline src/Spectre.Data.Tests/RoiReaderTests.cs]: Similarly as above, but for `Width`.

File: src/Spectre.Data.Tests/RoiWriterTests.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    class RoiWriterTests
+    {
+        [Test]
+        public void WriteRoi_writes_file_properly()
+        {
+            RoiWriter service = new RoiWriter();
+
+            service.RoiWriterTool(DataStub.WriteRoiDataset, DataStub.TestDirectoryPath);
+
+            RoiReader checkIfProperlyWrittenService = new RoiReader();
```
Comment (gmrukwa) [inline src/Spectre.Data.Tests/RoiWriterTests.cs]: Unnecessarily long name. `reader` maybe? :)

File: src/Spectre.Data/Datasets/RoiDictionary.cs [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+        /// <summary>
+        /// Tries to get specified value from list by name.
+        /// </summary>
+        /// <param name="name">The name.</param>
+        /// <returns>
+        /// Returns roi with specified name.
+        /// Returns null if no roi with specified name found.
+        /// </returns>
+        public Roi GetRoiOrDefault(string name)
```
Comment (gmrukwa) [inline src/Spectre.Data/Datasets/RoiDictionary.cs]: Can this function be simplified like:

`public Roi GetRoiOrDefault(string name) => _roiDataset.FirstOrDefault(r => r.Name ==name);`

I think it would suffice for doing exactly the same :)

File: src/Spectre.Data/Datasets/RoiDictionary.cs [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+        /// <summary>
+        /// Initializes the dictionary with specified path
+        /// and corresponding set of ROIs.
+        /// </summary>
+        /// <param name="path">The path.</param>
+        public void InitializeDictionary(string path)
+        {
+            _path = path;
+            _roiDataset = _roireader.GetAllRoisFromDirectory(path);
```
Comment (gmrukwa) [inline src/Spectre.Data/Datasets/RoiDictionary.cs]: What in the case, when there will be a lot of such files? I imagine situations when user in automated manner produces thousands of ROI, e.g. using DiviK. It would be good to read these lazily.

File: src/Spectre.Data/RoiIo/RoiReader.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        /// <summary>
+        /// Lists the rois from directory.
+        /// </summary>
+        /// <param name="path">The path.</param>
+        /// <returns>
+        /// All ROIs in specified directory.
+        /// </returns>
+        /// <exception cref="FileNotFoundException">No *.png files found in directory or subdirectories</exception>
+        public List<Roi> GetAllRoisFromDirectory(string path)
```
Comment (gmrukwa) [inline src/Spectre.Data/RoiIo/RoiReader.cs]: I think it would be useful, to have it splitted into two parts, one of which will work for single ROI. We have to be able to load ROIs in lazy way.

File: src/Spectre.Data/RoiIo/RoiReader.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+            var names = new List<string>();
+
+            var allfiles = System.IO.Directory.GetFiles(path, "*.png", SearchOption.TopDirectoryOnly);
+
+            names = allfiles.ToList();
+            names.Sort();
+
+            if (names.Count == 0)
+            {
+                throw new FileNotFoundException("No *.png files found in directory.");
```
Comment (gmrukwa) [inline src/Spectre.Data/RoiIo/RoiReader.cs]: I don't think it's a bad thing, that should not happen. It will happen for all newly uploaded files. Is this a case we shouldn't be able to handle? I don't think so.

File: src/Spectre.Data/RoiIo/RoiReader.cs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        /// <returns>
+        /// ROI dataset.
+        /// </returns>
+        public Roi RoiReaderTool(string path)
+        {
+            var roiConverter = new RoiConverter();
+            var bitmap = new Bitmap(path);
+
+            var roidataset = roiConverter.BitmapToRoi(bitmap, Path.GetFileNameWithoutExtension(path));
+            bitmap.Dispose();
```
Comment (gmrukwa) [inline src/Spectre.Data/RoiIo/RoiReader.cs]: Use `using` construct.

File: src/platform-components.sln [hunk 13]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
-VisualStudioVersion = 15.0.27004.2009
+VisualStudioVersion = 15.0.26730.3
```
Comment (gmrukwa) [inline src/platform-components.sln]: Unnecessary change.

File: src/platform-components.sln.orig [hunk 14]
```diff
@@ -0,0 +1,101 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
```
Comment (gmrukwa) [inline src/platform-components.sln.orig]: This file probably should not be here.

**Review by gmrukwa — CHANGES_REQUESTED** [review]
Really good piece of code. Appreciated! Some minor changes are needed to make it silky smooth. Can we also go for lazy loading of ROIs? This is the last one thing to be fixed, and we can go for merge. :)

File: src/Spectre.Data.Tests/RoiDictionaryTests.cs [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+
+        [Test]
+        public void LoadAllRois_loads_properly()
+        {
+            RoiDictionary roiDictionaryService = new RoiDictionary();
+            roiDictionaryService.SetDirectoryPath(DataStub.TestDirectoryPath);
+
+            roiDictionaryService.LoadAllRois();
+
+            var obtainedRoi = roiDictionaryService.GetRoiOrDefault("image1");
```
Comment (gmrukwa) [inline src/Spectre.Data.Tests/RoiDictionaryTests.cs]: Check [this](https://github.com/nunit/docs/wiki/Values-Attribute). It would be nice-to-have for your convenience next time. (I post this just for informational purposes - if you feel like applying it, that's fine, but it's not required for me).

File: src/Spectre.Data/Datasets/RoiDictionary.cs [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+    {
+        private List<Roi> _roiDataset = new List<Roi>();
+        private RoiReader _roireader;
+        private string _directoryPath;
+
+        /// <summary>
+        /// Sets the directory path.
+        /// </summary>
+        /// <param name="directoryPath">The directory path.</param>
+        public void SetDirectoryPath(string directoryPath)
```
Comment (gmrukwa) [inline src/Spectre.Data/Datasets/RoiDictionary.cs]: Let's keep it settable only through constructor, so this `RoiDictionary` is always bound to one directory. We don't want it to magically point to somewhere else after creation.

File: src/Spectre.Data/Datasets/RoiDictionary.cs [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+            _roireader = new RoiReader(_directoryPath);
+        }
+
+        /// <summary>
+        /// Loads all Rois from directory to the list.
+        /// All elements in the list will be overwritten.
+        /// </summary>
+        public void LoadAllRois()
+        {
+            _roiDataset = _roireader.GetAllRoisFromDirectory();
```
Comment (gmrukwa) [inline src/Spectre.Data/Datasets/RoiDictionary.cs]: It's good to cache these in memory, but why user can't also get them all at once (`return _roiDataset;`)? On the other hand, we should also check, whether `_roiDataset` hasn't been loaded yet - such situation is possible, and at this point we're reloading all of this from disk again... Even, if we consider that content of the disk could change, we should not reload everything! I'd even rather dropped caching of ROIs in RAM for the favor of having this lightweight.

File: src/Spectre.Data/RoiIo/RoiReader.cs [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+        /// All ROIs in specified directory.
+        /// </returns>
+        /// <exception cref="FileNotFoundException">No *.png files found in directory or subdirectories</exception>
+        public List<Roi> GetAllRoisFromDirectory()
+        {
+            var rois = new List<Roi>();
+
+            var names = GetAllNamesFromDirectory();
+
+            for (var listIterator = 0; listIterator < names.Count; listIterator++)
```
Comment (gmrukwa) [inline src/Spectre.Data/RoiIo/RoiReader.cs]: It's more readable to use LINQ in this place:
```
var rois = names.Select(name => GetSingleRoiFromDirectory(Path.Combine(_path, name))).ToList();
```

File: src/Spectre.Data/RoiIo/RoiReader.cs [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+        /// <summary>
+        /// Loads single ROI from specified folded.
+        /// </summary>
+        /// <param name="fileName">Name of the file with extension.</param>
+        /// <returns>
+        /// ROI dataset.
+        /// </returns>
+        public Roi GetSingleRoiFromDirectory(string fileName)
+        {
+            var roiConverter = new RoiConverter();
```
Comment (gmrukwa) [inline src/Spectre.Data/RoiIo/RoiReader.cs]: We're recreating this for each `Roi` read. Can we create it once, at creation of `RoiReader`?

**Review by gmrukwa — CHANGES_REQUESTED** [review]
I think that there are still few more "TODOs". I'll probably take a look at it.

**Comment by gmrukwa** [comment]
BTW. version number update is missing :D

File: src/Spectre.Data.Tests/RoiDictionaryTests.cs [hunk 20]
```diff
@@ -22,6 +22,7 @@
 using System.IO;
```
Comment (gmrukwa) [inline src/Spectre.Data.Tests/RoiDictionaryTests.cs]: Tests in this module became much cleaner. Good job!

File: src/Spectre.Data/RoiIo/RoiReader.cs [hunk 21]
```diff
@@ -80,6 +82,11 @@ public List<string> GetAllNamesFromDirectory()
 
             var allfiles = Directory.GetFiles(_path, "*.png", SearchOption.TopDirectoryOnly);
 
+            for (int iterator = 0; iterator < allfiles.Length; iterator++)
```
Comment (gmrukwa) [inline src/Spectre.Data/RoiIo/RoiReader.cs]: This could be done as above, using `Select`, but that's minor issue.

**Review by gmrukwa — APPROVED** [review]
Well done. Seems like we have it! 👍

**Review by Daroslaw — APPROVED** [review]
